imports/redolog_manager_factory.go

```diff
@@ -0,0 +1,157 @@
+}
+
+// Function definitio to store upsertbatch
+type BatchStoreFunc func(batch *memCom.UpsertBatch, redoFile int64, offset uint32, skipBackFillRows bool) error
+type newKafkaConsumerFunc func([]string) (sarama.Consumer, error)
+
+// NewRedologManagerFactory create ImportsFactory instance
+func NewRedologManagerFactory(c *common.ImportsConfig, diskStore diskstore.DiskStore, metaStore metastore.MetaStore) (*RedologManagerFactory, error) {
+	return NewKafkaRedologManagerFactory(c, diskStore, metaStore, nil)
+}
+
+func NewKafkaRedologManagerFactory(c *common.ImportsConfig, diskStore diskstore.DiskStore, metaStore metastore.MetaStore, consumer sarama.Consumer) (*RedologManagerFactory, error) {
+
+	if c != nil && c.Source == common.KafkaSoureOnly && len(c.KafkaConfig.Brokers) == 0 {
+		return nil, fmt.Errorf("No kafka broker info configured")
+	}
+
+	enableLocalRedoLog := true
+	if c != nil && c.Source == common.KafkaSoureOnly {
+		// we will turn on local redolog if import source is not Kafka only
+		enableLocalRedoLog = !c.RedoLog.Disabled
+	}
+
+	namespace := ""
+	if c != nil {
+		namespace = c.Namespace
+	}
+	factory := &RedologManagerFactory{
+		namespace:          namespace,
+		enableLocalRedoLog: enableLocalRedoLog,
+		diskStore:          diskStore,
+		managers:           make(map[string]map[int]RedologManager),
+		metaStore:          metaStore,
+	}
+
+	if c != nil {
+		if consumer == nil {
+			var err error
+			consumer, err = newKafkaConsumer(c.KafkaConfig.Brokers)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		factory.consumer = consumer
+	}
+
+	return factory, nil
+}
+
+// NewKafkaConsumer create kafka consumer
+func newKafkaConsumer(brokers []string) (sarama.Consumer, error) {
+	if len(brokers) > 0 {
+		return sarama.NewConsumer(brokers, sarama.NewConfig())
+	}
+	return nil, nil
+}
+
+// NewRedologManager create CompositeRedologManager on specified table/shard
+// each table/shard should only have one CompositeRedologManager
+func (f *RedologManagerFactory) NewRedologManager(table string, shard int, tableConfig *metaCom.TableConfig, storeFunc BatchStoreFunc) (RedologManager, error) {
+	utils.GetLogger().With("action", "ingestion", "table", table, "shard", shard).Info("Create Redolog Manager")
+	f.Lock()
+	defer f.Unlock()
+
+	var tableManager map[int]RedologManager
+	var ok bool
+	if tableManager, ok = f.managers[table]; !ok || tableManager == nil {
+		tableManager = make(map[int]RedologManager)
+		f.managers[table] = tableManager
+	}
+	manager, ok := tableManager[shard]
+	if ok && manager != nil {
+		fmt.Printf("error: %v\n", utils.StackError(nil, "error"))
+		return nil, fmt.Errorf("NewRedologManager for table: %s, shard: %d is already running", table, shard)
+	}
+
+	manager = NewCompositeRedologManager(table, shard, f, tableConfig, storeFunc)
+	tableManager[shard] = manager
+
+	return manager, nil
+}
+
+// Close one table shard Redolog manager
+func (f *RedologManagerFactory) Close(table string, shard int) {
```

Comment:
this function does not really close any redolog manager.
it is weird here that RedologManagerFactor close itself and then call factory.Close(table, shard) to remove itself.